Added postprocessing options for StreamInlet

RoboDoig opened the pull request:
Previously, StreamInlet creation used the default options which resulted in no online correction of LSL timestamps. This PR adds options for the various postprocessing options in LSL StreamInlets, allowing for e.g. online clock sync and timestamp monotonization.

RoboDoig committed:
Added postprocessing options to StreamInlet

RoboDoig committed:
Update LSL package for postprocessing options

glopesdev commented on src/Bonsai.Lsl/Bonsai.Lsl.csproj:
```diff
@@ -38,4 +38,10 @@
     <Content Include="..\liblsl\win-x64\bin\*.dll" PackagePath="runtimes\win-x64\native" CopyToOutputDirectory="PreserveNewest" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Compile Update="Native\LSL.cs">
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </Compile>
+  </ItemGroup>
+
```
Is this needed? It seems to copy the .cs source file itself to the output directory.

glopesdev commented on src/Bonsai.Lsl/StreamInlet.cs:
```diff
@@ -315,4 +321,40 @@ IObservable<Timestamped<TResult>> Generate<TResult>(Func<Native.StreamInlet, TRe
+    /// <summary>
+    /// Options for post-processing of samples for an LSL stream inlet.
+    /// </summary>
+    public enum ProcessingFlags
```
I would rename this enum to `ProcessingOptions` for consistency with LSL terminology.

glopesdev commented on src/Bonsai.Lsl/StreamInlet.cs:
```diff
@@ -57,6 +57,12 @@ public class StreamInlet : ZeroArgumentExpressionBuilder, INamedElement
+        /// Gets or sets a value specifying the postprocessing time-synchronisation options to use on an LSL stream inlet.
+        /// </summary>
+        [Description("Specifies the postprocessing options of an LSL stream inlet.")]
+        public ProcessingFlags ProcessingOptions { get; set; } = ProcessingFlags.All;
```
I would rename the type from `ProcessingFlags` to `ProcessingOptions`. It's fine to have the name of the property with the same name of the type.

RoboDoig replied:
Renamed in [0ec015a](https://github.com/emotional-cities/lsl/pull/5/commits/0ec015a7500057fc938eb7e0484649f20c1d4777)

glopesdev commented on src/Bonsai.Lsl/StreamInlet.cs:
```diff
@@ -315,4 +321,40 @@ IObservable<Timestamped<TResult>> Generate<TResult>(Func<Native.StreamInlet, TRe
+    /// <summary>
+    /// Options for post-processing of samples for an LSL stream inlet.
+    /// </summary>
+    public enum ProcessingFlags
```
Can you set multiple of these options at the same time? If so, consider adding the `[Flags]` attribute to the top of the enum declaration. That will make it so that in the property grid you can actually specify multiple options separated by a comma.

RoboDoig replied:
Yes you can, just tested on [aaaf8cc](https://github.com/emotional-cities/lsl/pull/5/commits/aaaf8ccea0a4493ac7b53dbd79c637551e0f7f5b)

glopesdev commented on src/Bonsai.Lsl/Bonsai.Lsl.csproj:
```diff
@@ -38,4 +38,10 @@
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </Compile>
+  </ItemGroup>
+
```
Is this needed? It seems to copy the .cs source file itself into the output directory.

RoboDoig replied:
Not needed, was just for testing. Removed in [0ec015a](https://github.com/emotional-cities/lsl/pull/5/commits/0ec015a7500057fc938eb7e0484649f20c1d4777)

glopesdev commented on src/Bonsai.Lsl/StreamInlet.cs:
```diff
@@ -215,7 +221,7 @@ static StreamInfo ResolveStream(string pred)
         }
 
-        static Native.StreamInlet CreateInlet(string name, string type)
+        static Native.StreamInlet CreateInlet(string name, string type, ProcessingFlags procFlags)
```
Avoid abbreviation here, and use either `processingFlags` or `processingOptions` as parameter name.

RoboDoig replied:
Changed in b7a21e20b9cc13eb9ade43d3157271b65ac68caa

glopesdev commented on src/Bonsai.Lsl/StreamInlet.cs:
```diff
@@ -57,6 +57,12 @@ public class StreamInlet : ZeroArgumentExpressionBuilder, INamedElement
+        /// <summary>
+        /// Gets or sets a value specifying the postprocessing time-synchronisation options to use on an LSL stream inlet.
+        /// </summary>
+        [Description("Specifies the postprocessing options of an LSL stream inlet.")]
```
For consistency, prefer US-english spelling in code and code comments (e.g. synchronization). Would update both descriptions for styling consistency with other properties and to match the XML comment (also there seems to be no hyphen in time-synchronization from looking at LSL docs):

Gets or sets a value specifying the postprocessing time synchronization options to use on the LSL stream.

"Specifies the postprocessing time synchronization options to use on the LSL stream."

RoboDoig replied:
Changed in b7a21e20b9cc13eb9ade43d3157271b65ac68caa

glopesdev commented on src/Bonsai.Lsl/StreamInlet.cs:
```diff
@@ -315,4 +321,40 @@ IObservable<Timestamped<TResult>> Generate<TResult>(Func<Native.StreamInlet, TRe
     }
+
+    /// <summary>
+    /// Options for post-processing of samples for an LSL stream inlet.
```
Specifies options for post-processing of samples for an LSL stream inlet.

RoboDoig replied:
Changed in b7a21e20b9cc13eb9ade43d3157271b65ac68caa

glopesdev commented on src/Bonsai.Lsl/StreamInlet.cs:
```diff
@@ -315,4 +321,40 @@ IObservable<Timestamped<TResult>> Generate<TResult>(Func<Native.StreamInlet, TRe
+    public enum ProcessingFlags
+    {
+        /// <summary>
+        /// No automatic post-processing. Gives ground-truth timestamps for manual post-processing.
```
No automatic post-processing. Provides ground truth timestamps for manual post-processing.

RoboDoig replied:
Changed in b7a21e20b9cc13eb9ade43d3157271b65ac68caa

glopesdev commented on src/Bonsai.Lsl/StreamInlet.cs:
```diff
@@ -315,4 +321,40 @@ IObservable<Timestamped<TResult>> Generate<TResult>(Func<Native.StreamInlet, TRe
+        None = processing_options_t.proc_none,
+
+        /// <summary>
+        /// Perform automatic clock-synchronisation.
```
Perform automatic clock synchronization.

RoboDoig replied:
Changed in b7a21e20b9cc13eb9ade43d3157271b65ac68caa

glopesdev commented on src/Bonsai.Lsl/StreamInlet.cs:
```diff
@@ -315,4 +321,40 @@ IObservable<Timestamped<TResult>> Generate<TResult>(Func<Native.StreamInlet, TRe
+        Clocksync = processing_options_t.proc_clocksync,
+
+        /// <summary>
+        /// Remove jitter from timestamps.
```
Remove random jitter from timestamps using a trend-adjusted smoothing algorithm.

reference: https://labstreaminglayer.readthedocs.io/info/time_synchronization.html

RoboDoig replied:
Changed in b7a21e20b9cc13eb9ade43d3157271b65ac68caa

glopesdev commented on src/Bonsai.Lsl/Bonsai.Lsl.csproj:
```diff
@@ -20,7 +20,7 @@
     <LangVersion>8.0</LangVersion>
     <Features>strict</Features>
-    <Version>0.2.0</Version>
+    <Version>0.2.1-rc1</Version>
```
Since we are adding new features, I would bump the minor version to `0.3.0`.

RoboDoig replied:
Changed in b7a21e20b9cc13eb9ade43d3157271b65ac68caa

RoboDoig committed:
Remove output copy and rename ProcessingFlags for consistency

RoboDoig committed:
Updating spelling, comments, version

- Removed abbreviations
- Check spelling consistency
- Clearer code comments, more consistent with LSL documentation
- Update minor version

RoboDoig committed:
Add flags attribute to ProcessingOptions, test on StreamInlet
